File: sdks/local-rp/csharp/src/LinkKeys.LocalRp/Revocation.cs
using LinkKeys.LocalRp.Wire;
using static LinkKeys.LocalRp.Wire.Types;

namespace LinkKeys.LocalRp;

/// <summary>
/// Sibling-signed key revocation certificate verification — mirrors
/// <c>crates/liblinkkeys/src/revocation.rs</c>. Only verification is ported here
/// (building/signing a revocation certificate is a domain-admin/server-side operation,
/// out of scope for a local-RP SDK); this SDK verifies revocation certificates fetched
/// alongside domain keys so it can drop a key a quorum-verified sibling revocation
/// targets.
/// </summary>
public static class Revocation
{
    /// <summary>Minimum number of distinct sibling signatures required to revoke a key.</summary>
    public const int Quorum = 2;

    /// <summary>Domain-separation tag/version for the signed revocation payload.</summary>
    private const string Tag = "linkkeys-key-revocation-v1";

    /// <summary>
    /// The canonical signed bytes: the tag, the target key id + fingerprint, the
    /// revocation instant, and the signing sibling's domain (bound per-signature to stop
    /// cross-domain reuse of a signature). This is the OLDER house tuple pattern — a
    /// five-element array with the domain-separation tag first — NOT the two-element
    /// <c>CBOR([context, payload])</c> envelope framing the four local-RP structures use.
    /// </summary>
    internal static byte[] RevocationPayload(string targetKeyId, string targetFingerprint, string revokedAt, string signingDomain) =>
        Cbor.Encode(Cbor.Tuple(
            Cbor.VTextOf(Tag), Cbor.VTextOf(targetKeyId), Cbor.VTextOf(targetFingerprint), Cbor.VTextOf(revokedAt),
            Cbor.VTextOf(signingDomain)));

    /// <summary>
    /// Verify a revocation certificate against a domain's public key set. Requires at
    /// least <see cref="Quorum"/> DISTINCT signing keys of <paramref name="domain"/>,
    /// each currently valid and NOT the target key, to have signed the canonical payload.
    /// </summary>
    public static void VerifyRevocationCertificate(RevocationCertificate cert, IReadOnlyList<DomainPublicKey> domainKeys, string domain)
    {
        int counted = CountValidSigners(cert, domainKeys, domain);
        if (counted < Quorum)
        {
            throw new RevocationError(counted, Quorum);
        }
    }

    /// <summary>
    /// The number of distinct, currently-valid, non-self, correctly-signed sibling
    /// signatures the certificate carries for <paramref name="domain"/>. Exposed
    /// (internal) so conformance tests can pinpoint exactly which filtering rule an
    /// implementation got wrong, per <c>revocations.json</c>'s
    /// <c>expected_counted_signers</c> field; <see cref="VerifyRevocationCertificate"/>
    /// is the only entry point production code should call.
    /// </summary>
    internal static int CountValidSigners(RevocationCertificate cert, IReadOnlyList<DomainPublicKey> domainKeys, string domain)
    {
        var validSigners = new HashSet<string>();

        foreach (var sig in cert.Signatures)
        {
            // A key can never authorize its own revocation.
            if (sig.SignedByKeyId == cert.TargetKeyId)
            {
                continue;
            }

            // The signature must be bound to this domain.
            if (sig.Domain != domain)
            {
                continue;
            }

            var key = domainKeys.FirstOrDefault(k => k.KeyId == sig.SignedByKeyId);
            if (key is null)
            {
                continue;
            }

            // Only a currently-valid signing key counts toward the quorum.
            try
            {
                LocalRp.CheckSigningKeyValid(key);
            }
            catch (LocalRpError)
            {
                continue;
            }

            var payload = RevocationPayload(cert.TargetKeyId, cert.TargetFingerprint, cert.RevokedAt, sig.Domain);
            if (key.Algorithm == "ed25519" && key.PublicKey.Length == 32 && Crypto.Crypto.VerifyEd25519(payload, sig.Signature, key.PublicKey))
            {
                validSigners.Add(sig.SignedByKeyId);
            }
        }

        return validSigners.Count;
    }
}

File: sdks/local-rp/java/src/main/java/community/catalyst/linkkeys/localrp/Revocation.java
package community.catalyst.linkkeys.localrp;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import community.catalyst.linkkeys.localrp.crypto.Crypto;
import community.catalyst.linkkeys.localrp.wire.Cbor;
import community.catalyst.linkkeys.localrp.wire.Types.ClaimSignature;
import community.catalyst.linkkeys.localrp.wire.Types.DomainPublicKey;
import community.catalyst.linkkeys.localrp.wire.Types.RevocationCertificate;

/**
 * Sibling-signed key revocation certificate verification &mdash; mirrors
 * {@code crates/liblinkkeys/src/revocation.rs} / {@code sdks/local-rp/go/revocation.go}.
 * Only verification is ported here (building/signing a revocation
 * certificate is a domain-admin/server-side operation, out of scope for a
 * local-RP SDK); this SDK verifies revocation certificates fetched alongside
 * domain keys so it can drop a key a quorum-verified sibling revocation
 * targets.
 */
public final class Revocation {
    private Revocation() {}

    /** Minimum number of distinct sibling signatures required to revoke a key. */
    public static final int QUORUM = 2;

    /** Domain-separation tag/version for the signed revocation payload. */
    private static final String TAG = "linkkeys-key-revocation-v1alpha";

    /**
     * The canonical signed bytes: the tag, the target key id + fingerprint,
     * the revocation instant, and the signing sibling's domain (bound
     * per-signature to stop cross-domain reuse of a signature). This is the
     * OLDER house tuple pattern &mdash; a five-element array with the
     * domain-separation tag first &mdash; NOT the two-element
     * {@code CBOR([context, payload])} envelope framing the four local-RP
     * structures use.
     */
    static byte[] revocationPayload(String targetKeyId, String targetFingerprint, String revokedAt, String signingDomain) {
        return Cbor.encode(Cbor.tuple(
                Cbor.vtext(TAG),
                Cbor.vtext(targetKeyId),
                Cbor.vtext(targetFingerprint),
                Cbor.vtext(revokedAt),
                Cbor.vtext(signingDomain)));
    }

    /**
     * Verify a revocation certificate against a domain's public key set.
     * Requires at least {@link #QUORUM} DISTINCT signing keys of
     * {@code domain}, each currently valid and NOT the target key, to have
     * signed the canonical payload.
     */
    public static void verifyRevocationCertificate(
            RevocationCertificate cert, List<DomainPublicKey> domainKeys, String domain) {
        int counted = countValidSigners(cert, domainKeys, domain);
        if (counted < QUORUM) {
            throw new RevocationError(counted, QUORUM);
        }
    }

    /**
     * The number of distinct, currently-valid, non-self, correctly-signed
     * sibling signatures the certificate carries for {@code domain}. Exposed
     * (package-private) so conformance tests can pinpoint exactly which
     * filtering rule an implementation got wrong, per
     * {@code revocations.json}'s {@code expected_counted_signers} field;
     * {@link #verifyRevocationCertificate} is the only entry point
     * production code should call.
     */
    static int countValidSigners(RevocationCertificate cert, List<DomainPublicKey> domainKeys, String domain) {
        Set<String> validSigners = new HashSet<>();

        for (ClaimSignature sig : cert.signatures()) {
            // A key can never authorize its own revocation.
            if (sig.signedByKeyId().equals(cert.targetKeyId())) {
                continue;
            }
            // The signature must be bound to this domain.
            if (!sig.domain().equals(domain)) {
                continue;
            }
            DomainPublicKey key = domainKeys.stream()
                    .filter(k -> k.keyId().equals(sig.signedByKeyId()))
                    .findFirst()
                    .orElse(null);
            if (key == null) {
                continue;
            }
            // Only a currently-valid signing key counts toward the quorum.
            try {
                LocalRp.checkSigningKeyValid(key);
            } catch (LocalRpError e) {
                continue;
            }
            byte[] payload = revocationPayload(cert.targetKeyId(), cert.targetFingerprint(), cert.revokedAt(), sig.domain());
            if ("ed25519".equals(key.algorithm())
                    && key.publicKey().length == 32
                    && Crypto.verifyEd25519(payload, sig.signature(), key.publicKey())) {
                validSigners.add(sig.signedByKeyId());
            }
        }

        return validSigners.size();
    }
}
